### ISSUE TITLE
About the output of radius in Jetson

### ISSUE THREAD

**Issue by crea397** [rusty1s/pytorch_cluster#109]
This issue is based on https://github.com/rusty1s/pytorch_geometric/discussions/2758.

The following is the log of the test failure when running `python setup.py test` on Jetson Xavier NX.
Sometimes the test succeeded and sometimes it failed.

Please check the logs.
Thanks!

## Log of failures

$ python setup.py test
running pytest
Searching for pytest-cov
Best match: pytest-cov 2.12.1
Processing pytest_cov-2.12.1-py3.6.egg

Using /home/nx/workspace/pytorch_cluster/.eggs/pytest_cov-2.12.1-py3.6.egg
Searching for coverage>=5.2.1
Best match: coverage 5.6b1
Processing coverage-5.6b1-py3.6-linux-aarch64.egg

Using /home/nx/workspace/pytorch_cluster/.eggs/coverage-5.6b1-py3.6-linux-aarch64.egg
running egg_info
writing torch_cluster.egg-info/PKG-INFO
writing dependency_links to torch_cluster.egg-info/dependency_links.txt
writing requirements to torch_cluster.egg-info/requires.txt
writing top-level names to torch_cluster.egg-info/top_level.txt
reading manifest file 'torch_cluster.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'torch_cluster.egg-info/SOURCES.txt'
running build_ext
copying build/lib.linux-aarch64-3.6/torch_cluster/_rw.so -> torch_cluster
copying build/lib.linux-aarch64-3.6/torch_cluster/_sampler.so -> torch_cluster
copying build/lib.linux-aarch64-3.6/torch_cluster/_radius.so -> torch_cluster
copying build/lib.linux-aarch64-3.6/torch_cluster/_fps.so -> torch_cluster
copying build/lib.linux-aarch64-3.6/torch_cluster/_knn.so -> torch_cluster
copying build/lib.linux-aarch64-3.6/torch_cluster/_grid.so -> torch_cluster
copying build/lib.linux-aarch64-3.6/torch_cluster/_graclus.so -> torch_cluster
copying build/lib.linux-aarch64-3.6/torch_cluster/_nearest.so -> torch_cluster
copying build/lib.linux-aarch64-3.6/torch_cluster/_version.so -> torch_cluster
============================================================================================================================ test session starts =============================================================================================================================
platform linux -- Python 3.6.9, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
rootdir: /home/nx/workspace/pytorch_cluster, configfile: setup.cfg
plugins: cov-2.12.1
collected 85 items

test/test_fps.py ......
test/test_graclus.py ................
test/test_grid.py ................................
test/test_knn.py ............
test/test_nearest.py F...
test/test_radius.py ............
test/test_rw.py ..
test/test_sampler.py .

================================================================================================================================== FAILURES ==================================================================================================================================
________________________________________________________________________________________________________________________ test_nearest[dtype0-device0] ________________________________________________________________________________________________________________________

dtype = torch.float32, device = device(type='cpu')

    @pytest.mark.parametrize('dtype,device', product(grad_dtypes, devices))
    def test_nearest(dtype, device):
        x = tensor([
            [-1, -1],
            [-1, +1],
            [+1, +1],
            [+1, -1],
            [-2, -2],
            [-2, +2],
            [+2, +2],
            [+2, -2],
        ], dtype, device)
        y = tensor([
            [-1, 0],
            [+1, 0],
            [-2, 0],
            [+2, 0],
        ], dtype, device)

        batch_x = tensor([0, 0, 0, 0, 1, 1, 1, 1], torch.long, device)
        batch_y = tensor([0, 0, 1, 1], torch.long, device)

        out = nearest(x, y, batch_x, batch_y)
>       assert out.tolist() == [0, 0, 1, 1, 2, 2, 3, 3]
E       assert [0, 0, 0, 0, 2, 2, ...] == [0, 0, 1, 1, 2, 2, ...]
E         At index 2 diff: 0 != 1
E         Use -v to get the full diff

test/test_nearest.py:33: AssertionError

----------- coverage: platform linux, python 3.6.9-final-0 -----------
Name                        Stmts   Miss  Cover
__________________________________________________
torch_cluster/__init__.py      15      0   100%
torch_cluster/fps.py            2      0   100%
torch_cluster/graclus.py        2      0   100%
torch_cluster/grid.py           2      0   100%
torch_cluster/knn.py            2      0   100%
torch_cluster/nearest.py       21      0   100%
torch_cluster/radius.py         2      0   100%
torch_cluster/rw.py             3      0   100%
torch_cluster/sampler.py        1      0   100%
__________________________________________________
TOTAL                          50      0   100%

========================================================================================================================== short test summary info ===========================================================================================================================
FAILED test/test_nearest.py::test_nearest[dtype0-device0] - assert [0, 0, 0, 0, 2, 2, ...] == [0, 0, 1, 1, 2, 2, ...]
======================================================================================================================= 1 failed, 84 passed in 12.90s ========================================================================================================================

**Comment by rusty1s** [comment]
Thanks a lot, it looks like the `radius` test succeeds, but the nearest implementation has some issues (related to `scipy`?). We have multiple options to proceed:
1. Can you add your point cloud as an additional test (similar to [`test_radius_graph_large`](https://github.com/rusty1s/pytorch_cluster/blob/master/test/test_radius.py#L62-L73)) and see if it succeeds?
2. We probably need to debug [`radius_cuda.cu`](https://github.com/rusty1s/pytorch_cluster/blob/master/csrc/cuda/radius_cuda.cu) in order to idenfity the issues on Jetson. You may wanna check if decreasing `THREADS` to something smaller (e.g., `256` or `512`) fixes the issue for you.

**Comment by crea397** [comment]
> Can you add your point cloud as an additional test (similar to test_radius_graph_large) and see if it succeeds?

I tested it with some of my point clouds and the following code.
The code is almost the same as `test_radius_graph_large`.

```Python
import torch
import scipy
from torch_cluster import radius_graph
import numpy as np
import open3d as o3d


def to_set(edge_index):
    return set([(i, j) for i, j in edge_index.t().tolist()])


def my_test_radius_graph_large():
    # x = torch.randn(1000, 3)

    pcd = o3d.io.read_point_cloud('./xxx.pcd')  # Read Point Cloud
    pcd_np_arr = np.asarray(pcd.points).astype(np.float32)  # Point Cloud 2 Numpy
    x = torch.from_numpy(pcd_np_arr)  # Numpy 2 Tensor

    edge_index = radius_graph(x, r=0.5, flow='target_to_source', loop=True,
                              max_num_neighbors=max_num_neighbors, num_workers=6)

    tree = scipy.spatial.cKDTree(x.numpy())
    col = tree.query_ball_point(x.cpu(), r=0.5)
    truth = set([(i, j) for i, ns in enumerate(col) for j in ns])

    assert to_set(edge_index) == truth
```

I changed the value of `max_num_neighbors` and ran the test for several of my point clouds.
The following table shows the results of the test.

### Jetson
|Point Cloud  |64  |128  |256  |512  |1024  |2000  |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|default x  |○ or ×  |○ or ×  |○ or ×  |○ or ×  |○ or ×  |○ or ×  |
|chair  |×  |○  |○  |○  |○  |○  |
|rabbit  |×  |○  |○  |○  |○  |○  |
|car  |×  |×  |×  |×  |×  |×  |
|person  |×  |×  |×  |×  |×  |×  |

### Ubuntu
|Point Cloud  |64  |128  |256  |512  |1024  |2000  |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|default x  |○  |○  |○  |○  |○  |○  |
|chair  |×  |○  |○  |○  |○  |○  |
|rabbit  |×  |○  |○  |○  |○  |○  |
|car  |×  |×  |○  |○  |○  |○  |
|person  |×  |×  |×  |×  |○  |○  |

### Point Clound
|Point Cloud  |size  |
|:---:|:---:|
|default x  |(1000, 3)  |
|chair  |(111, 3)  |
|rabbit  |(113, 3)  |
|car  |(1315, 3)  |
|person  |(1059, 3)  |

---

> We probably need to debug radius_cuda.cu in order to idenfity the issues on Jetson. You may wanna check if decreasing THREADS to something smaller (e.g., 256 or 512) fixes the issue for you.

First I changed the value of `THREADS` to `256` and `512` and ran `python setup.py test` again. 
As a result of running the test, the test sometimes failed and sometimes succeeded whether the value of `THREADS` was `256` or `512`.

I have a question about your second piece of advice.
I have `pytorch_cluster` already installed to use `pytorch_geometric`, and I think the value of `THREADS` is fixed at `1024`.
How can I change the value of `THREADS` and do the inference in Jetson?

**Comment by rusty1s** [comment]
Thanks for the experiments. So it indeed looks like Ubuntu machine produces the correct results, while Jetson does not for `car` and `person`.

Keep in mind that you need to install `torch-cluster` locally on your machine in order to modify the number of `THREADS`. For example:
```
# Uninstall all previous versions of torch-cluster
git clone torch-cluster
cd pytorch_cluster
python setup.py develop
# Test the code
# Change `THREADS`
python setup.py develop  # Re-compile with modified THREADS
```

**Comment by crea397** [comment]
Thank you for the instructions on how to change the value of `THREADS`.
I changed the value of `THREADS` in Jetson's `radius_cuda.cu` to `256` and `512` and ran the inference in Jetson, but the issues were not fixed.

**Comment by rusty1s** [comment]
Thanks. I see that your test actually only runs on CPU. Does the same issue happens when running on GPU?

I would really love to dig into this and fix this for Jetson (there might be some inherent issues with Jetson if it fails on both CPU and GPU), but I sadly do not have access to it.

**Comment by crea397** [comment]
Thank you very much for your kind attention to my problem.

I am sorry, I thought I was running the tests (by tests do you mean inference in Jetson?) on the GPU.

Why did you think you were running the tests on the CPU?
Also, how can I run the tests on the GPU?

**Comment by rusty1s** [comment]
I see. The test weren't actually executed on GPU. This is now fixed. You can see in [the test code](https://github.com/rusty1s/pytorch_cluster/blob/master/test/test_radius.py#L62), that the test will be executed across different dtypes and devices, e.g., for CPU and GPU.

**Comment by crea397** [comment]
Thanks for the fix. I changed `my_test_radius_graph_large` above, referring to the [test code](https://github.com/rusty1s/pytorch_cluster/blob/master/test/test_radius.py#L62) that you fixed.

The following table shows the results of running the test with the modified `my_test_radius_graph_large`.
The results are the same as the Ubuntu results described above.

### Jetson
|Point Cloud  |64  |128  |256  |512  |1024  |2000  |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|default x  |○  |○  |○  |○  |○  |○  |
|chair  |×  |○  |○  |○  |○  |○  |
|rabbit  |×  |○  |○  |○  |○  |○  |
|car  |×  |×  |○  |○  |○  |○  |
|person  |×  |×  |×  |×  |○  |○  |

From this result, I thought there was a problem with my code.
There was actually a problem with my code and I was able to solve my problem by fixing the code.
Here is a report on what I fixed.

---

First, I used the GPU after loading the point cloud and creating the `Data` object. The source code is below.

### Creating the Data object
```Python
# Data.y
label = torch.tensor([class_num], dtype=torch.int64)

# Data.pos
pcd = o3d.io.read_point_cloud(pcd_path)  # Read Point Cloud
pcd_np_arr = np.asarray(pcd.points).astype(np.float32)  # Point Cloud 2 Numpy
pcd_tensor_arr = torch.from_numpy(pcd_np_arr)  # Numpy 2 Tensor

# Data object
data = Data(pos=pcd_tensor_arr, y=label)
```

### inference
```Python
model.eval()

for data in loader:
    print(data.pos.device, data.pos.dtype)  # --> cpu, torch.float32
    data = data.to(device)  # Transfer data object to GPU
    print(data.pos.device, data.pos.dtype)  # --> cuda:0, torch.float32

    with torch.no_grad():
        pred = model(data).max(1)[1]
```

After applying `to(device)` before inference, `data.pos.device` seems to have changed from `cpu` to `cuda:0`.
However, this approach seems to be the cause of my problem.

In fact, if I set up the `pos` to use the GPU before creating the `Data` object, as in the code below, the inference accuracy in Jetson did not decrease.

### Creating the Data object
```Python
# Data.y
label = torch.tensor([class_num], dtype=torch.int64, device=device)

# Data.pos
pcd = o3d.io.read_point_cloud(pcd_path)  # Read Point Cloud
pcd_np_arr = np.asarray(pcd.points).astype(np.float32)  # Point Cloud 2 Numpy
pcd_tensor_arr = torch.from_numpy(pcd_np_arr)  # Numpy 2 Tensor
pcd_tensor_arr = pcd_tensor_arr.to(device)

# Data object
data = Data(pos=pcd_tensor_arr, y=label)
```

I'm very sorry that there was a problem with my code for this issue.
I am curious as to why this issue occurred only with Jetson.

**Comment by rusty1s** [comment]
So the issue was than Jetson inference was running on CPU, while the Ubuntu one was running on GPU?

**Comment by crea397** [comment]
Yes, it seems that in Jetson, my unmodified code above is actually inferring on the CPU.

**Comment by rusty1s** [comment]
Ok, I see. There is a slight difference in how the `radius` op will operate on CPU and GPU. On CPU, it will pick the nearest neighbors in case the actual number of neighbors is larger than `max_num_neighbors`, will on GPU it will pick neighbors at random. However, with `knn` both computations should yield identical results.

**Comment by crea397** [comment]
I see that there is such a difference in the behavior of `radius` between CPU and GPU.
I am deeply grateful to you for your polite response to this issue.
